### PR TITLE
Add function editDistanceUTF8

### PR DESCRIPTION
**Pull request by liuneng1994**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add new function `editDistanceUTF8`, which calculates the [edit distance](https://en.wikipedia.org/wiki/Edit_distance) between two UTF8 strings.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit d642002ecd463bc30ca98cbfba905a9ca581f54c with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65269/d642002ecd463bc30ca98cbfba905a9ca581f54c/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/02884_string_distance_function.sql
Comment (superdiaodiao) [inline tests/queries/0_stateless/02884_string_distance_function.sql:55]: How about add a test in the '-- non-const arguments' part above, just after the line 33?
Like the function stringJaccardIndexUTF8 did.

Comment (liuneng1994) [reply tests/queries/0_stateless/02884_string_distance_function.sql:55]: ok

File: src/Functions/FunctionsStringDistance.cpp
Comment (alexkats) [inline src/Functions/FunctionsStringDistance.cpp:553]: nit: please add indentation here

File: tests/queries/0_stateless/02884_string_distance_function.sql
Comment (alexkats) [inline tests/queries/0_stateless/02884_string_distance_function.sql:29]: Let's add one more test from the example here as well: `('我是谁', '我是我')`

File: src/Functions/FunctionsStringDistance.cpp
Comment (alexkats) [inline src/Functions/FunctionsStringDistance.cpp:229]: Looks very similar to parsing a string in `ByteJaccardIndexImpl`, do you think we can deduplicate it a little bit?

**Review by alexkats — CHANGES_REQUESTED** [review]
Thanks for the changes, in general looks good, I left a couple of comments. Also let's squash some commits before merging it to master to have a cleaner history

File: src/Functions/FunctionsStringDistance.cpp
Comment (alexkats) [inline src/Functions/FunctionsStringDistance.cpp:254]: Looks very similar to parsing a string in `ByteJaccardIndexImpl`, do you think we can deduplicate it a little bit?

Comment (liuneng1994) [reply src/Functions/FunctionsStringDistance.cpp:254]: ok, let me try

**Comment by Avogar** [comment]
Test `03036_dynamic_read_subcolumns` was failing in master. It's already fixed so please update the branch with recent changes in master to avoid its failures in the CI.

**Review by alexkats — APPROVED** [review]
Thanks, lgtm

**Pull request merged**
